src/main/java/address/ui/Ui.java

```diff
@@ -1,39 +1,445 @@
+    private static final String FXML_ACTIVITY_HISTORY = "/view/ActivityHistory.fxml";
+    private static final String FXML_STATUS_BAR_FOOTER = "/view/StatusBarFooter.fxml";
+    private static final String FXML_TAG_EDIT_DIALOG = "/view/TagEditDialog.fxml";
+    private static final String FXML_TAG_LIST = "/view/TagList.fxml";
+    private static final String FXML_BIRTHDAY_STATISTICS = "/view/BirthdayStatistics.fxml";
+    private static final String FXML_TAG_SELECTION_EDIT_DIALOG = "/view/TagSelectionEditDialog.fxml";
+
+    private static final String ICON_EDIT = "/images/edit.png";
+    private static final String ICON_CALENDAR = "/images/calendar.png";
+    private static final String ICON_INFO = "/images/info_icon.png";
+
+
+    private MainApp mainApp; //TODO: remove this back link to higher level class
+
+    //Links to internal non-UI components
+    private ModelManager modelManager;
 
-    public Ui(MainApp mainApp, ModelManager modelManager, Config config, UserPrefs pref){
-        mainController = new MainController(mainApp, modelManager, config, pref);
-        this.pref = pref;
+    //Objects containing parameters for controlling App behavior
+    private Config config;
+    private UserPrefs prefs;
+
+    //Main Window of the app
+    private MainWindow mainWindow;
+
+    private BrowserManager browserManager;
+
+    //TODO: move these inside MainWindowUiPart (similar to PersonListPanel)
+    private StatusBarHeaderController statusBarHeaderController;
+    private StatusBarFooterController statusBarFooterController;
+
+    //TODO: See if these can be pushed down to respective UI parts
+    private UnmodifiableObservableList<ReadOnlyViewablePerson> personList;
+    private final ObservableList<SingleTargetCommandResultEvent> finishedCommandResults;
+
+    {
+        finishedCommandResults = FXCollections.observableArrayList();
     }
 
-    public void start(Stage primaryStage) {
-        mainController.start(primaryStage);
+
+    public Ui(MainApp mainApp, ModelManager modelManager, Config config, UserPrefs prefs) {
+        this.mainApp = mainApp;
+        this.modelManager = modelManager;
+        this.config = config;
+        this.prefs = prefs;
+        this.personList = modelManager.getAllViewablePersonsReadOnly();
+        this.browserManager = new BrowserManager(personList, config.getBrowserNoOfPages(), config.getBrowserType());
+        this.browserManager.initBrowser();
+        EventManager.getInstance().registerHandler(this);
     }
 
-    public void showAlertDialogAndWait(Alert.AlertType alertType, String alertTitle, String headerText, String contentText) {
-        mainController.showAlertDialogAndWait(alertType, alertTitle, headerText, contentText);
+    public void start(Stage primaryStage) {
+        try {
+            logger.info("Starting main UI.");
+
+            mainWindow = MainWindow.load(primaryStage, config, prefs, mainApp, this, modelManager);
```

Comment:
Moved the construction logic completely from the call site and put it in the respective class. This is an example of the clean call site.